src/array.jl

```diff
@@ -552,18 +552,66 @@ function copyto!(dest::CatArrOrSub{T, N, R}, dstart::Integer,
     dest
 end
 
-copyto!(dest::CatArrOrSub, src::CatArrOrSub) =
+function copyto!(dest::CatArrOrSub{T1, N, R}, dstart::Integer,
+                 src::AbstractArray{T2,N}, sstart::Integer,
+                 n::Integer) where {T1, T2, N, R}
+    n == 0 && return dest
+    n < 0 && throw(ArgumentError(string("tried to copy n=", n, " elements, but n should be nonnegative")))
+    destinds, srcinds = LinearIndices(dest), LinearIndices(src)
+    (checkbounds(Bool, destinds, dstart) && checkbounds(Bool, destinds, dstart+n-1)) || throw(BoundsError(dest, dstart:dstart+n-1))
+    (checkbounds(Bool, srcinds, sstart)  && checkbounds(Bool, srcinds, sstart+n-1))  || throw(BoundsError(src,  sstart:sstart+n-1))
+    srclevs = DataAPI.refpool(src)
+    srcrefs = DataAPI.refarray(src)
+    # Fast path only supports refs which can be used to index
+    # into vector mapping from source levels to destination levels
+    if !(srclevs isa AbstractVector) || !(srcrefs isa AbstractArray{<:Integer})
+        return invoke(copyto!, Tuple{AbstractArray, Integer, AbstractArray, Integer, Integer},
+                      dest, dstart, src, sstart, n)
+    end
+    checkbounds(dest, axes(src)...)
```

Comment:
I am not sure that this check makes sense. It is not done in Base AFAICT. You already checked if there enough space in `destinds` and `srcinds` above - right?

Reply:
Good catch. That's probably a leftover from a point where I intended to implement only `copyto!(x, y)`.